### ISSUE TITLE
Are the changes in CardDAV-sync in 2021.02.2?

### ISSUE THREAD

**Issue by ignoreigor** [tine20/tine20#7300]
Hi,

I am syncing my addressbooks not only to mobile devices but also use the FRITZ!Box' ability to sync addressbooks via CardDAV and transfer it to a Fritz!Fon. Since the update to Tine 2.0, version 2021.02.2, my contacts changed. Instead of the formerly synced personal contacts I am getting contacts from the "internal contacts (shared)" folder. 

Setting up a new addressbook, I only get this folder's contacts, no matter what CalDAV-URI I use. From to root to the concrete folder's URI, I always get contacts from the "shared addressbook"-folder.

I know, that you don't know the Fritz!Box' behaviour, I will do some tests with another DAV-sync client and have a look in the logs.

Best regards,

**Comment by lab-at-nohl** [comment]
Hi ignoreigor, 

thanks for the tip, didn't know that AVM makes it possible finally. Works very well, besides my phone does not use the phone book from Fritzbox as default. 

You can always select the exact Addressbook by its URL, right click and select properties... Otherwise "shared" might just be the first book in the index listing (this is the case here). 

Johannes

**Comment by ignoreigor** [comment]
Hi Johannes,

> thanks for the tip, didn't know that AVM makes it possible finally. Works very well, besides my phone does not use the phone book from Fritzbox as default.

Yes, it's a nice side-effect to use your Tine 2.0 addressbooks.

> 
> You can always select the exact Addressbook by its URL, right click and select properties... Otherwise "shared" might just be the first book in the index listing (this is the case here).

I know. I did so. And it worked, but since updating contacts in the FB's phonebook changed and even if I set it up again, I always get the shared contacts.

Well, I think I should have another look on it.

**Comment by ignoreigor** [comment]
Okay, here are some facts. Tried with two different users, say USERA and USERB, to sync each personal addressbook to FritzBox.

Request URI für personal addressbook arrives server, okay FB is sending the complete String. Calling this URI in a browser shows the correct folder contents.
`^[[0m078c51 - ^[[1m^[[32mUSERA^[[0m - ^[[1m^[[32m2021-04-25T23:50:17+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/addressbooks/4dc71a8074e8e5b3bd47a1de2e161154cc7e0afe/0ea9c9e6deb4d2b89fa31ef9cf79d39f6ef23e7d/ PID: 988515)
`
Got this error... don't know where this principal should be. Don't know if there was this error before.
`^[[0m078c51 - ^[[1m^[[32mUSERA^[[0m - ^[[1m^[[36m2021-04-25T23:50:17+02:00 NOTICE (5): Tinebase_WebDav_PrincipalBackend::getPrincipalByPath::176 Group/list principal does not exist: 99b8d030802e872c8fd01072f99e255c14f3431e`

Now it tries one level higher...
`^[[0ma9f7a7 - ^[[1m^[[32mUSERA^[[0m - ^[[1m^[[32m2021-04-25T23:50:17+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/principals/users/4dc71a8074e8e5b3bd47a1de2e161154cc7e0afe/ PID: 986358)`

And got same error again:
`^[[0ma9f7a7 - ^[[1m^[[32mUSERA^[[0m - ^[[1m^[[36m2021-04-25T23:50:17+02:00 NOTICE (5): Tinebase_WebDav_PrincipalBackend::getPrincipalByPath::176 Group/list principal does not exist: 99b8d030802e872c8fd01072f99e255c14f3431e
`
Now the shared addressbook is looked up. Same error follows, but this folders contacts are synced.
`^[[0m081676 - ^[[1m^[[32mUSERA^[[0m - ^[[1m^[[32m2021-04-25T23:50:19+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/addressbooks/4dc71a8074e8e5b3bd47a1de2e161154cc7e0afe/1/ PID: 988855)`

Now let USERB do the same (personal container's id is shorter):
`
^[[0m54ea7e - ^[[1m^[[32mUSERB^[[0m - ^[[1m^[[32m2021-04-26T00:44:08+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/addressbooks/803021c9a571e250024307b8b85537301c8a3240/2/ PID: 987402)`

Got two errors for two principals
```
^[[0m54ea7e - ^[[1m^[[32mUSERB^[[0m - ^[[1m^[[36m2021-04-26T00:44:08+02:00 NOTICE (5): Tinebase_WebDav_PrincipalBackend::getPrincipalByPath::176 Group/list principal does not exist: 99b8d030802e872c8fd01072f99e255c14f3431e
^[[0m54ea7e - ^[[1m^[[32mUSERB^[[0m - ^[[1m^[[36m2021-04-26T00:44:08+02:00 NOTICE (5): Tinebase_WebDav_PrincipalBackend::getPrincipalByPath::176 Group/list principal does not exist: ea7dd50cfa543317e19c815deabcf05eb45f5843
```
Ask one level higher
`^[[0m4a4a8a - ^[[1m^[[32mUSERB^[[0m - ^[[1m^[[32m2021-04-26T00:44:09+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/addressbooks/803021c9a571e250024307b8b85537301c8a3240/ PID: 964599)
`
Two principal errors again...

And here the shared addressbook folder's id of USERB, that is synced to F!B
`^[[0m7d40f4 - ^[[1m^[[32mUSERB^[[0m - ^[[1m^[[32m2021-04-26T00:44:10+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/addressbooks/803021c9a571e250024307b8b85537301c8a3240/1/ PID: 988681)`

Any ideas what might going wrong?

**Comment by lab-at-nohl** [comment]
Who/what is the principal `99b8d030802e872c8fd01072f99e255c14f3431e`?

I would guess that this is a deleted user. However, it might have remained in access control list or so. Maybe removal was unclean. But it would be worth to examine this. 

Can look in your browser who/what is exposed there? Simply enter the URL and provide user credentials. Then have a look, personal folders, shared folders and ...?

Also can you look in the database where this string occurs?

**Comment by ignoreigor** [comment]
Did some filtering of the dump, because there were many appearances in tine20_timemachine_modlog and -got it.
 
> Who/what is the principal `99b8d030802e872c8fd01072f99e255c14f3431e`?

It's the primary key of table tine20_addressbook_lists, the entries of group "Users" and "Administrators", both groups that USERA (only Users) und USERB are members of.

> +------------------------------------------+--------------+----------------+----------------------------------+-------+-------+------------------------------------------+---------------------+------------------------------------------+---------------------+------------+------------+--------------+-------------+-----+--------+
> | id                                       | container_id | name           | description                      | email | type  | created_by                               | creation_time       | last_modified_by                         | last_modified_time  | is_deleted | deleted_by | deleted_time | list_type   | seq | xprops |
> +------------------------------------------+--------------+----------------+----------------------------------+-------+-------+------------------------------------------+---------------------+------------------------------------------+---------------------+------------+------------+--------------+-------------+-----+--------+
> | 99b8d030802e872c8fd01072f99e255c14f3431e | 1            | Users          | Group of user accounts           | NULL  | group | 5d6aa4c9bc8929e6410f1440ee2bfe36072f7c94 | 2011-12-08 12:31:14 | NULL                                     | NULL                |          0 | NULL       | NULL         | NULL        |   0 | NULL   |
> | ea7dd50cfa543317e19c815deabcf05eb45f5843 | 1            | Administrators | Group of administrative accounts | NULL  | group | NULL                                     | NULL                | 5d6aa4c9bc8929e6410f1440ee2bfe36072f7c94 | 2012-01-02 10:18:43 |          0 | NULL       | NULL         | NULL        |   0 | NULL   |
> +------------------------------------------+--------------+----------------+----------------------------------+-------+-------+------------------------------------------+---------------------+------------------------------------------+---------------------+------------+------------+--------------+-------------+-----+--------+

> I would guess that this is a deleted user. However, it might have remained in access control list or so. Maybe removal was unclean. But it would be worth to examine this.

at least the Groups are not deleted.

> 
> Can look in your browser who/what is exposed there? Simply enter the URL and provide user credentials. Then have a look, personal folders, shared folders and ...?

I did as I wrote in the 4. entry: "Calling this URI in a browser shows the correct folder contents." - I can call the folder and after logging in, I got the vcf-files of the desired folder, I want to sync.

Syncing this folder with the FB did already work, but the content of the addressbook changed since I did the last update and is no longer possible to get this contacts. I get no failure, but the content of the shared addressbook instead.

I will do some very dangerous: I will try the code of the last release on the current database and look what will happen... Of course I backuped the DB before.

'll keep you informed.

**Comment by lab-at-nohl** [comment]
> both groups that USERA (only Users) und USERB are members of.

I do not think the _principal_ is a regular group. It looks like a contact list (table: _addressbook_lists), which is sometimes called group as well. Do you have the App CoreData installed? You can find those in `Application Data/Addressbook/Groups`. 

For any further details a log file, level DEBUG, is needed.

**Comment by ignoreigor** [comment]
Hi Johannes,

thank you for your effords, CoreData is installed in version 13.0.
There a two groups, replicators and one mail-distribution group, I created. I see both groups in Addressbook-application -> Groups and in CoreData-application -> Application Data/Addressbook/Groups as well. 

I stumbled over the addressbooks-part in the tablename, too. The principals, primary keys of tine20_addressbook_lists are foreign key in tine20_groups and points to group Administrator and Users. So here is the relation to systemgroups. Sorry, I missed to look for.

Because this behaviour now happens in the former version, I will try to test it without caching or clean my redis before and see what'll happen. You can even see the changed behaviour in the nginx-access logs before.

I raised my loglevel from 3, where I can see almost nothing, to 6, where I found the entries I posted earlier. I can go up to 7 (DEBUG), but I should stop tine-cronjobs before an have a lot of work (especially to find)  and wipe personal data out of the debuglog. Then I can attach some logs.

**Comment by lab-at-nohl** [comment]
Of course, take your time. I realized that my installation give this NOTICE, too. That's why I followed the issue further. In my case the error is raised here: [Tinebase/WebDav/PrincipalBackend.php](https://github.com/tine20/tine20/blob/2021.02.2/tine20/Tinebase/WebDav/PrincipalBackend.php#L176)

```php
                    $filter = new Addressbook_Model_ListFilter(array(
                        array(
                            'field' => 'type',
                            'operator' => 'equals',
                            'value' => Addressbook_Model_List::LISTTYPE_GROUP
                        ),
                        array(
                            'field' => 'id',
                            'operator' => 'equals',
                            'value' => $id
                        ),
                    ));

                    $list = Addressbook_Controller_List::getInstance()->search($filter)->getFirstRecord();

                    if (!$list) {
                        if (Tinebase_Core::isLogLevel(Zend_Log::NOTICE)) {
                            Tinebase_Core::getLogger()->notice(
                                __METHOD__ . '::' . __LINE__ . ' Group/list principal does not exist: ' . $id);
                        }
                        return null;
                    }
```

To me it seems harmless. But obviously $id is not empty, however, the - existing - group/principal is not found. Hence, we have a bug here, I guess...

Please have a look to logging configuration. You can select a username to log, if you don't want to change your settings.

**Comment by ignoreigor** [comment]
Hi,

here is a complete debug log, did some cleaning sed's. Please give me a hint, if you still see sensible data (other than uids). I turned off cron jobs, you see only one sync attemp from the Fritz!Box
```

80:117871 - USERA - 2021-05-06T21:49:44+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/2/ PID: 2257142)
253:dba5bf - USERA - 2021-05-06T21:49:44+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/tine20/principals/users/803021c9a571e250024307b8b85537301c8a3240/ PID: 2604003)
395:26b91b - USERA - 2021-05-06T21:49:44+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/ PID: 2257147)
585:de0d8e - USERA - 2021-05-06T21:49:46+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/1/ PID: 2257149)
797:999de2 - USERA - 2021-05-06T21:49:46+02:00 INFO (6): Tinebase_Server_WebDAV::handle::194 Starting to handle WebDAV request ( requestUri:/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/1/ PID: 2257140)
```
In Line 80 you see the URI of the folder, I want to sync (/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/2/), I inserted in the Fritz!Box. In line 797 you can see the call of the folder whose contacts are finally synced to the Fritz!Box - the shared addressbook. Starting with line 866 you see the contact's data.

[debug.log.gz](https://github.com/tine20/tine20/files/6447715/debug.log.gz)

I wouldn't be that surprised if it didn't suddenly change its behavior about a month ago, after working for months, and now contacts are synced from the wrong address book

**Comment by lab-at-nohl** [comment]
Although I cannot tell what is going on, I'd like to share my findings. First of all, your _Fritzbox_ ignores that you provided a fixed folder `803021c9a571e250024307b8b85537301c8a3240/2/`. Instead the _Fritzbox_ asks several PROPFINDS and then decides to go with **another folder than you have named explicitly by its url**.

### The PROPFINDS in detail:

1) **current-user-principal** returns `/tine20/principals/users/803021c9a571e250024307b8b85537301c8a3240/` (in my case it returns calendar proxy entries which are empty)
2) **addressbook-home-set** also returns `/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/` (all personal folders of the user)
3) **displayname** finally returns several folders
- **Lastname, Firstname** -> `/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/`
- **Internal Contacts (shared)** -> `/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/1/`
- **Tine 2.0 Admin Accounts persönliches Adressbuch** -> `/tine20/addressbooks/803021c9a571e250024307b8b85537301c8a3240/2/`

### Analysis

I cannot tell what _Fritzbox_ is doing. It seems reasonable not to go with the user default because it is for calendar. And the call for the home addressbook returns the base folder, which is empty. So in the end, the displayname list is used. I guess the first subfolder is chosen because _Fritzbox_ thinks it is the most specific... However, it is different to the specific folder you have given. This is the main issue and related to _Fritzbox_, imho. If subfolders cannot be selected in the UI the _Fritzbox_ should include all folders combined, not only the first in the list...

**I can reproduce** this behavior. My _Fritzbox_ ALWAYS returns the first personal folder, too. The order is simply set by alphabet. 

Btw, it is only due to the Admin account that a shared folder gets synced. Also with my Tine 2.0 installation, all shared folders get listed as "personal" subfolders (probably because Admin is the owner). To have the shared folders listed as personal folders might be a Tine 2.0 bug - not sure.

**Comment by ignoreigor** [comment]
Hi Johannes,

thanks for your look and your testing. I now believe that it is a FritzBox-issue not to use the selected folder. But I wonder why it worked for a long time. 

I have some apache access-logs of the time when it worked:


Shared folders appear as personal folder for any user, that has access to. 

But maybe your alphabetical approach is not so bad. I will try to rename the folder to change the order and'll see, 
There was no update of my Fritz!Box, but I switched from apache to nginx. Maybe the directory-listing is influenced, though the content is not static but script-generated.
